refactor: Use JSON files for TraceExporter client initialization

kmimu opened the pull request:
This PR addresses the issue mentioned in Issue #404 by implementing [the suggested Option 2](https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/issues/404#issuecomment-1669892639)

- Refactored TraceExporter client initialization
- Replaced direct proto file references with JSON files
- Converted proto files to JSON using the `npm run compile-protos` command

Checks:
- [x] Make sure to open an issue as a [bug/issue/feature](https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #404

MichaelDoyle commented:
@aabmass would you be able to take a look at this?

aabmass commented on packages/opentelemetry-cloud-trace-exporter/package.json:
```diff
@@ -10,11 +10,12 @@
     "fix": "gts fix",
-    "pretest": "npm run compile"
+    "pretest": "npm run compile",
+    "compile-protos": "node compile-protos.js"
```
Can you add this to run during `precompile` step?

kmimu replied:
I tested a few scenarios, and it turns out that if we move the JSON file directly into the build directory and remove `/protos/protos.json`, the reference from `trace.ts` results in an error.

However, during my testing, I realized that we don’t need to run `cp -r protos build` because the JSON file is already included in the build due to the include option in the tsconfig.json. Therefore, I have removed that command since it was redundant. → [1558ee4](https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/pull/741/commits/1558ee4d3b7cff4e214b621dd3e1adccef8585b4)

If you know of any alternative ways to handle this without removing `protos/protos.json`, I’d appreciate your idea!

aabmass replied:
I guess we could simply add it to .gitiginore.

However I think we what have here is fine–I doubt the generated code will change often anyway. Thank you for looking into it.

aabmass commented on packages/opentelemetry-cloud-trace-exporter/compile-protos.js:
```diff
@@ -0,0 +1,60 @@
+    await pbjsMain(pbjsArgs);
+    console.log('Proto JSON file generated successfully.');
+  } catch (error) {
+    console.error('Error generating Proto JSON file:', error);
```
I think we would want this script to fail. Can you re-throw the error or just not catch the error?

kmimu replied:
fixed this in [cf66489](https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/pull/741/commits/cf664897f7b2b496cd8c2f14810cca861bbb49cb)

aabmass reviewed (commented):
Thank you for the PR! I left a few comments regarding using typescript and a few questions.

aabmass commented:
/gcbrun

kmimu commented on packages/opentelemetry-cloud-trace-exporter/src/trace.ts:
```diff
@@ -20,14 +20,14 @@ import {
 import {getReadableSpanTransformer} from './transform';
 import {TraceService, NamedSpans} from './types';
 import {VERSION} from './version';
+const protoJson = require('../protos/protos.json');
```
fixed this in [0fb04c4](https://github.com/GoogleCloudPlatform/opentelemetry-operations-js/pull/741/commits/0fb04c4f32218ee2bc9d9a99e881f7d430f3e2f2)

kmimu commented on packages/opentelemetry-cloud-trace-exporter/tsconfig.json:
```diff
@@ -2,7 +2,8 @@
     "rootDir": ".",
-    "outDir": "build"
+    "outDir": "build",
+    "resolveJsonModule": true,
```
No, users of this library will not necessarily need to turn on resolveJsonModule in their TS config unless they are directly importing JSON files in their own project. If library uses JSON internally and is already compiled, the consumers will interact with the built JavaScript or type definition files.

For example, if you add a line like this (and compile):
```
diff --git a/packages/opentelemetry-cloud-trace-exporter/src/trace.ts b/packages/opentelemetry-cloud-trace-exporter/src/trace.ts
index 2b73a36..21e52ee 100644
--- a/packages/opentelemetry-cloud-trace-exporter/src/trace.ts
+++ b/packages/opentelemetry-cloud-trace-exporter/src/trace.ts
@@ -30,6 +30,9 @@ import {VERSION} from './version';
 import * as protoJson from '../protos/protos.json';
 
+export const value = protoJson.nested.google.nested.devtools.nested.cloudtrace.nested.v2.nested.TraceService.methods;
```

You can see in trace.d.ts that the type information can be retrieved without directly referencing the JSON in the user's code:
```
export declare const value: {
    BatchWriteSpans: {
        requestType: string;
        responseType: string;
        options: {
            "(google.api.http).post": string;
            "(google.api.http).body": string;
            "(google.api.method_signature)": string;
....
```

aabmass replied:
Very clear, thank you 🙂

kmimu commented:
@aabmass 
Thanks for your initial review! I’ve addressed all the comments, and I’d really appreciate it if you could take another look when you have a chance.

aabmass commented:
/gcbrun

aabmass commented:
/gcbrun

aabmass commented:
/gcbrun
